src/compose/config.py

```diff
@@ -0,0 +1,195 @@
+# SPDX-FileCopyrightText: 2023 CERN (home.cern)
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""Load, parse and validate configuration."""
+
+import os
+import subprocess  # noqa: S404
+from datetime import date
+from http import HTTPMethod, HTTPStatus
+from logging import debug, info
+from tempfile import TemporaryDirectory
+from typing import Annotated, Literal, Optional, TextIO, Union
+from urllib import request
+from urllib.error import URLError
+from urllib.parse import urlparse
+
+import yaml
+from pydantic import AfterValidator, BaseModel, ValidationError
+
+
+class ConfigError(Exception):
+    """Failed to load, parse or validate configuration."""
+
+
+def is_reachable(url: str) -> str:
+    """
+    Check if a string is a reachable URL.
+
+    Parameters:
+        url: URL string.
+
+    Returns:
+        URL string.
+
+    Raises:
+        ValueError: if the URL is not reachable.
+    """
+    req = request.Request(url, method=HTTPMethod.HEAD)
+    debug('Checking if URL is reachable: {0}.'.format(url))
+    try:
+        with request.urlopen(req) as response:  # noqa: S310
+            if response.status != HTTPStatus.OK:
+                msg = 'URL is unreachable: {0}:\n↳ Status Code: {1}'
+                raise ValueError(msg.format(url, response.status))
+    except URLError as error:
+        raise ValueError('URL is unreachable: {0}:\n↳ {1}'.format(url, error))
+    return url
+
+
+ReachableURL = Annotated[str, AfterValidator(is_reachable)]
```

Comment:
is it possible to encapsulate ReachableURL and is_reachable() in a class?

Reply:
Yes, it will bloated the code a bit but I just moved it to a class.
More organized this way - https://github.com/vascoguita/ohwr.org/blob/1a326251468747589da150d48eec91e7832fb840/src/compose/url.py